Ultra Low Power deepsleep

hallard opened the pull request:
- Added Ultra Low Power mode (30uA sleep mode)
- Wake by LoRa module, avoid watchdog wake every 8s
- Added new method to read battery voltage, needed if VCC<2.5V
- Updated documentation
- Bug fix with button wake
- Press button more than 2s to enable 1minute window to upload new sketch when in deepsleep

hallard committed:
bugfix and features

- Added Hardware Reset of lora module at startup
- Fixed push button duration bug
- Set IRQ var to be volatile

hallard committed:
Added Low Power Example

hallard committed:
Creation

hallard committed:
Added Lora Module Sleep Mode

hallard committed:
Fixed link

hallard committed:
bump to version 2.0.3

hallard committed:
force cast of CONFIG_INTERVAL #define

hallard committed:
Only TTN Core team allowed to increase version

hallard committed:
Cosmetic

hallard committed:
cosmetic formating

hallard committed:
Added missing include

hallard committed:
Requires TheThingsNode library 2.5.6

hallard committed:
Added option to wake by lora module

Moved #include definition here
Added wakeStatus to wakeCallback to get what waked us
Added pointer to TheThingsNetwork object to enable Lora module
sleep/wake up

hallard committed:
added option to wake by lora module

removed #include definition (in .h now)
Added wakeStatus bit field to wakeCallback to get what waked us
Added pointer to TheThingsNetwork object to enable Lora module
sleep/wake up
Added IRQ for INT2 (Lora Serial Module)
Added new configInterval method
Optimized Light Sensor
Changed deepSleep to support wake by lora module

hallard committed:
Fix Lora Module Sleep

hallard committed:
Fix IRQ problem on Serial Wake not detected

hallard committed:
Added new sleep mode with LoRa module

hallard committed:
cosmetic

hallard committed:
Updated to new Ultra Low Power Sleep Mode (with LoRa module wake)

hallard committed:
Detailled Low Power Mode

hallard committed:
Merge remote-tracking branch 'upstream/master'

hallard committed:
Added getVCC() using with ADC Low Noise Feature

hallard committed:
Get VCC of ATMega with getVCC()

hallard committed:
Fixed button release bug

hallard committed:
send all VDD/VCC/VRN values

hallard committed:
Added The Thing Network Library 2.5.7

hallard committed:
use of sprintf_p and some cosmetics

hallard committed:
Disable USB Regulator on deep sleeep

hallard committed:
Added Ultra Low Power mode

hallard committed:
send wakeSatus to intervall callback

hallard committed:
Low Power optimization and bug fixes

Refactored Interrupt
Refactored button management
improove deep sleep mode
disable USB VREG on sleep

hallard committed:
Optimized Low Power

hallard committed:
Updated for Ultra Low Power

hallard committed:
fix Link

hallard committed:
fix code dispkay

hallard committed:
Updated for Ultra Low Power

hallard committed:
secrure printf and set channel 10 for button duration

hallard committed:
updated onInterval method()

hallard committed:
fixed wake with wakeReason parameter

hallard committed:
fixed call of wake and interval

hallard committed:
fix var name error

johanstokking commented on docs/TheThingsNode.md:
```diff
@@ -38,23 +38,76 @@ This will:
-void wake() {
-    node->setColor(TTN_GREEN);
+void wake(uint8_t wakeReason) {
+  debugSerial.print(F("Wake Reason:0x"));
```
Spacing; `Reason: 0x`

johanstokking commented on docs/TheThingsNode.md:
```diff
@@ -38,23 +38,76 @@ This will:
+  debugSerial.print(F("Wake Reason:0x"));
+  debugSerial.println(wakeReason,HEX);
+
+  if (wakeReason&TTN_WAKE_WATCHDOG) {
```
Spacing; `wakeReason & TTN_WAKE_WATCHDOG`

johanstokking commented on docs/TheThingsNode.md:
```diff
@@ -38,23 +38,76 @@ This will:
-    node->setColor(TTN_GREEN);
+void wake(uint8_t wakeReason) {
+  debugSerial.print(F("Wake Reason:0x"));
+  debugSerial.println(wakeReason,HEX);
```
Spacing; `wakeReason, HEX`

johanstokking commented on docs/TheThingsNode.md:
````diff
@@ -97,31 +148,37 @@ void setup() {
 ```
 
 - `bool enabled`: Enable or disable the interval callback. Enabled automatically by `onInterval()`, but you can use this method to temporarily disable it. Defaults to `false`.
+- `TheThingsNetwork * pttn`: This enable the interval callback but in this mode, the interval is passed to RN2483 or RN2903 module (this is why we need to pass pointer to object) with the command `sys sleep ms` and then it's the LoRa module that wake up the node. This is the most advanced Low Power Mode. In this mode, the watchdog is disabled and consuption again reduced.
````
`consuMption`

johanstokking commented on examples/BatteryMonitorLPP/BatteryMonitorLPP.ino:
```diff
@@ -62,89 +131,127 @@ void sleep()
-  debugSerial.print(timepressed);
-  debugSerial.println(F(" ms"));
+  snprintf_P(buf, sizeof(buf), PSTR("-- SEND: BUTTON %d ms"), timepressed);
+  debugSerial.println(buf);
```
Build the string here with print/println calls for consistency

johanstokking commented on examples/BatteryMonitorLPP/BatteryMonitorLPP.ino:
```diff
@@ -62,89 +131,127 @@ void sleep()
+  if (value) {
+    snprintf_P(buf, sizeof(buf), PSTR("Btn:\t %dms"), value);
+    debugSerial.println(buf);
+	  lpp.addAnalogInput(10, value/1000.0);
```
Indentation + spacing

johanstokking commented on examples/BatteryMonitorLPP/BatteryMonitorLPP.ino:
```diff
@@ -62,89 +131,127 @@ void sleep()
-    debugSerial.println(F("ms"));
-    lpp.addAnalogInput(7, value / 1000.0);
+  if (value) {
+    snprintf_P(buf, sizeof(buf), PSTR("Btn:\t %dms"), value);
```
Built string. Also, what is `Btn`

johanstokking commented on examples/BatteryMonitorLPP/BatteryMonitorLPP.ino:
```diff
@@ -62,89 +131,127 @@ void sleep()
 
+  node->setColor(TTN_BLUE);
+  // Send data
+  //ttn.sendBytes(lpp.getBuffer(), lpp.getSize(), port);
```
Remove line

johanstokking commented on examples/CayenneLPP/CayenneLPP.ino:
```diff
@@ -79,24 +112,46 @@ void sleep()
+  if (bat <= 3300) 
+  {
+    bat = node->getVcc();
+  }
```
Shouldn't `getBattery()` do this?

hallard replied:
Yes, but `getBattery()` fails under 2.5V because ADC reference is 2.5V so in this case we use another technique that use the 1.1V ADC reference voltage and measure the Atmel VCC pin
see https://www.thethingsnetwork.org/forum/t/the-things-node-whats-this-3-2ma-sleep-mode-set-to-30ua-now/12521/74

johanstokking replied:
Yes but I mean should't `getBattery()` implement this fallback itself?

hallard replied:
Absolutely, it's in my list ;-)
But I'm under testing the difference for now in sketch BatteryMonitorLPP to see difference, so I need to have your original battery measurement to see differences.

johanstokking commented on examples/CayenneLPP/CayenneLPP.ino:
```diff
@@ -1,74 +1,107 @@
+// This sketch use advanced Ultra Low Power techniques
```
This sketch is mostly about CayenneLPP

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -4,67 +4,31 @@
 {
   uint8_t trigger = getPinChangeInterruptTrigger(digitalPinToPCINT(TTN_ACCELEROMETER_INT2));
+  // Reset according bits
+  wakeStatus &= ~ (TTN_WAKE_MOTION_START|TTN_WAKE_MOTION_STOP);
```
Spacing

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -183,54 +186,84 @@ void TheThingsNode::loop()
-
-    while (!TTN_BUTTON_PRESS && !TTN_BUTTON_RELEASE && !TTN_MOTION_START && !TTN_MOTION_STOP && !TTN_TEMPERATURE && TTN_INTERVAL < this->intervalMs)
+    // Loop until pseudo wake event (because we're not sleeping) or interval
+    while (!(wakeStatus&TTN_WAKE_ANY) && TTN_INTERVAL < this->intervalMs)
```
Spacing + consistent order (i.e. constant on right side)

hallard replied:
Not sure to understand what you mean, would you provide an example?

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -293,16 +326,23 @@ void TheThingsNode::showStatus()
 void TheThingsNode::configInterval(bool enabled, uint32_t ms)
 {
   this->intervalMs = ms;
```
Shouldn't `pttn` be cleared here as well? Like L341

hallard replied:
Nope because it's done just after by the call to 
` configInterval(enabled);`
but I admit it's not read friendly, so I changed this for consistency

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  adcIRQCnt = 0;
+
+  // We want to have a interrupt when the conversion is done
+  ADCSRA |= _BV( ADIE );
```
Spacing

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  // 8 samples (we don't take the 1st one)
+  do {
+    // Enable Noise Reduction Sleep Mode
+    set_sleep_mode( SLEEP_MODE_ADC );
```
Spacing

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  ADCSRA |= _BV(ADSC); 
+  
+  // wait for first dummy conversion
+  while (bit_is_set(ADCSRA,ADSC));
```
Use `{` and `}`; the `;` only is quite unreadable

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+      cli();
+    }
+    // Check is done with interrupts disabled to avoid a race condition
+    while (bit_is_set(ADCSRA,ADSC));
```
Spacing

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  ADCSRA &= ~ _BV( ADIE );
+  
+  // Return the average divided by 8 (8 samples)
+  return ( average ? sum >> 3 : sum );
```
Spacing

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+Output  : value readed
+Comments: ADC Channel input is modified
+====================================================================== */
+uint16_t TheThingsNode::getVcc() 
```
Make it `getVCC`?

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+Input   : -
+Output  : value readed
+Comments: ADC Channel input is modified
+====================================================================== */
```
Remove comment block for consistency

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  value = readADCLowNoise(true);
+
+  // Vcc reference in millivolts
+  vcc =  ( 1023L * 1100L) / value ; 
```
Spacing

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  
+  // Operating range of ATMega
+  if (vcc < 1800 ) vcc = 1800 ;
+  if (vcc > 5500 ) vcc = 5500 ;
```
Spacing and indentation

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -691,6 +746,102 @@ void TheThingsNode::setColor(ttn_color color)
+  if (vcc > 5500 ) vcc = 5500 ;
+    
+  // Vcc in millivolts
+  return ( vcc ); 
```
Why parentheses?

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -748,14 +903,13 @@ TheThingsNode::TheThingsNode()
 
-  // hardware reset of LoRa module, so module is reset on sketch upload 
-  #ifdef TTN_LORA_RESET
+  // reset RN2483 module, this allow to reset module on sketch upload also
```
`RN2xx3`

johanstokking commented on src/TheThingsNode.cpp:
```diff
@@ -890,20 +1044,57 @@ void TheThingsNode::WDT_stop()
+    cli();  
+    bitSet(EIFR,INTF2); // clear any pending interrupts for serial RX pin (INT2 D0)
+    sei();
+
```
Remove empty line

johanstokking reviewed (changes requested):
Nice, thanks very much! Mostly code style issues.

Also looping in @laurensslats for testing

hallard commented:
@johanstokking you want me to remove any use of printf, sprintf and snprintf, correct?

hallard committed:
renamed getVcc to getVCC

hallard committed:
renamed getVcc to getVCC and cosmetic

hallard committed:
renamed getVcc to getVCC

hallard committed:
renamed getVcc to getVCC

hallard committed:
costemic requested by johan

hallard committed:
removed all printf reference

hallard committed:
removed printf reference requested by johan

hallard commented:
Requested changes are done

johanstokking commented:
@hallard we try to keep dynamic memory allocations to a minimum, at least that's a goal!

hallard committed:
Clarification about sensor working mode

see

https://www.thethingsnetwork.org/forum/t/the-things-node-whats-this-3-2ma-sleep-mode-set-to-30ua-now/12521/160

hallard committed:
Updated temp sensor to Low Power

hallard commented:
removing printf is not a bad idea, I think it also save same pgm space (as soon as it's not used anywhere else in any lib)

hallard committed:
Added delay on wake if temp sensor was in power down mode

hallard committed:
fix push button debouncing in IRQ

hallard committed:
Disable Sleep mode in USB

hallard committed:
Disabled sleep mode in USB

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/TheThingsNetwork/arduino-node-lib?pullRequest=15) <br/>All committers have signed the CLA.
